From 7604d4ee3d93e502ed89a0f09503f5968e1234e1 Mon Sep 17 00:00:00 2001
From: Thecrazyskull <Thecrazyskull@users.noreply.github.com>
Date: Tue, 29 Sep 2020 14:53:57 +0000
Subject: [PATCH 2/3] PowerManagerService: Allow to light up buttons only when
 pressed

Author: Anas Karbila <anaskarbila@gmail.com>
Date:   Sat Jun 3 03:21:32 2017 +0200

    PowerManagerService: Allow to light up buttons only when pressed

     * Right now capactive, lit hardware keys are being
       lit every time you either touch them or the screen.

       But some devices handle this differently on stock:

       Display touch => buttons not lit
       Buttons touch => buttons lit

     * Thus, add a setting in order to allow the user
       to choose the preferred behavior.

    Change-Id: I35ac71a8274568901f962c9692788d1c682a98dd

Author: Corinna Vinschen <xda@vinschen.de>
Date:   Sun Aug 6 15:05:54 2017 +0200

    PowerManagerService: fix HW button illumination timeout

    Change I35ac71a8274568901f962c9692788d1c682a98dd, introducing hardware
    button backlight on button keypress only, also introduced a bug:

    When touching a button and then performing display activity while
    the buttons are still on, the buttons would keep lightened up until
    the next user interaction, potentially only switched off at the next
    screen off timeout.  Also, the buttons were not illuminated on
    device wakeup.

    This patch fixes it, together with another, long-standing problem:

    When touching a hardware button, nextTimeout was set to
    now + mButtonTimeout, even if mButtonTimeout is longer than the timeout
    determined by the screen off timeout.  To wit, if screen timeout is set
    to 15 secs, but button timeout to values > 15 secs.

    Change-Id: I8a56f1d1e0138c38ed6fe294e4816a9f7f744f1e
    Signed-off-by: Corinna Vinschen <xda@vinschen.de>

Change-Id: I5b486e65a5b7d9d16590941df0af4d9c604dedc4
---
 core/java/android/provider/Settings.java           |  9 +++++
 .../android/server/power/PowerManagerService.java  | 41 +++++++++++++++++++---
 2 files changed, 45 insertions(+), 5 deletions(-)

diff --git a/core/java/android/provider/Settings.java b/core/java/android/provider/Settings.java
index ad409d6a3cc..2c390c4e620 100644
--- a/core/java/android/provider/Settings.java
+++ b/core/java/android/provider/Settings.java
@@ -4803,6 +4803,14 @@ public final class Settings {
          */
         public static final String BUTTON_BRIGHTNESS = "button_brightness";
 
+        /**
+         * Whether the button backlight is only lit when pressed (and not when screen is touched)
+         * The value is boolean (1 or 0).
+         * @hide
+         */
+        public static final String BUTTON_BACKLIGHT_ONLY_WHEN_PRESSED =
+                "button_backlight_only_when_pressed";
+
         /**
          * Whether or not to enable multiple audio focus.
          * When enabled, requires more management by user over application playback activity,
@@ -5238,6 +5246,7 @@ public final class Settings {
             PRIVATE_SETTINGS.add(DISPLAY_COLOR_MODE);
             PRIVATE_SETTINGS.add(BUTTON_BACKLIGHT_TIMEOUT);
             PRIVATE_SETTINGS.add(BUTTON_BRIGHTNESS);
+            PRIVATE_SETTINGS.add(BUTTON_BACKLIGHT_ONLY_WHEN_PRESSED);
             PRIVATE_SETTINGS.add(VOLUME_PANEL_ON_LEFT);
             PRIVATE_SETTINGS.add(POCKET_JUDGE);
             PRIVATE_SETTINGS.add(DOUBLE_TAP_SLEEP_GESTURE);
diff --git a/services/core/java/com/android/server/power/PowerManagerService.java b/services/core/java/com/android/server/power/PowerManagerService.java
index 0c4b3eb1469..c3bb5fe92a2 100644
--- a/services/core/java/com/android/server/power/PowerManagerService.java
+++ b/services/core/java/com/android/server/power/PowerManagerService.java
@@ -278,6 +278,10 @@ public final class PowerManagerService extends SystemService
     private InattentiveSleepWarningController mInattentiveSleepWarningOverlayController;
     private final AmbientDisplaySuppressionController mAmbientDisplaySuppressionController;
 
+    private boolean mButtonPressed;
+    private boolean mButtonOn;
+    private boolean mButtonLightOnKeypressOnly;
+
     private final Object mLock = LockGuard.installNewLock(LockGuard.INDEX_POWER);
 
     // A bitfield that indicates what parts of the power state have
@@ -329,6 +333,7 @@ public final class PowerManagerService extends SystemService
     private int mLastSleepReason;
 
     // Timestamp of the last call to user activity.
+    private long mLastButtonActivityTime;
     private long mLastUserActivityTime;
     private long mLastUserActivityTimeNoChangeLights;
 
@@ -1167,6 +1172,9 @@ public final class PowerManagerService extends SystemService
         resolver.registerContentObserver(Settings.System.getUriFor(
                 Settings.System.BUTTON_BACKLIGHT_TIMEOUT),
                 false, mSettingsObserver, UserHandle.USER_ALL);
+        resolver.registerContentObserver(Settings.System.getUriFor(
+                Settings.System.BUTTON_BACKLIGHT_ONLY_WHEN_PRESSED),
+                false, mSettingsObserver, UserHandle.USER_ALL);
 
         IVrManager vrManager = IVrManager.Stub.asInterface(getBinderService(Context.VR_SERVICE));
         if (vrManager != null) {
@@ -1300,6 +1308,9 @@ public final class PowerManagerService extends SystemService
         mButtonBrightness = Settings.System.getFloatForUser(resolver,
                 Settings.System.BUTTON_BRIGHTNESS, mButtonBrightnessSettingDefault,
                 UserHandle.USER_CURRENT);
+         mButtonLightOnKeypressOnly = Settings.System.getIntForUser(resolver,
+                Settings.System.BUTTON_BACKLIGHT_ONLY_WHEN_PRESSED,
+                0, UserHandle.USER_CURRENT) == 1;
 
         mDirty |= DIRTY_SETTINGS;
     }
@@ -1679,6 +1690,12 @@ public final class PowerManagerService extends SystemService
                 }
             } else {
                 if (eventTime > mLastUserActivityTime) {
+                    mButtonPressed = event == PowerManager.USER_ACTIVITY_EVENT_BUTTON;
+                    if ((mButtonLightOnKeypressOnly && mButtonPressed)
+                            || eventTime == mLastWakeTime) {
+                        mButtonPressed = true;
+                        mLastButtonActivityTime = eventTime;
+                    }
                     mLastUserActivityTime = eventTime;
                     mDirty |= DIRTY_USER_ACTIVITY;
                     if (event == PowerManager.USER_ACTIVITY_EVENT_BUTTON) {
@@ -2332,14 +2349,27 @@ public final class PowerManagerService extends SystemService
                                 buttonBrightness = mButtonBrightness;
                             }
 
+                            mLastButtonActivityTime = mButtonLightOnKeypressOnly ?
+                                    mLastButtonActivityTime : mLastUserActivityTime;
                             if (mButtonTimeout != 0 &&
-                                    now > mLastUserActivityTime + mButtonTimeout) {
+                                    now > mLastButtonActivityTime + mButtonTimeout) {
                                 mButtonsLight.setBrightness(PowerManager.BRIGHTNESS_OFF_FLOAT);
+                                mButtonOn = false;
                             } else {
-                                mButtonsLight.setBrightness(buttonBrightness);
-                                if (buttonBrightness != PowerManager.BRIGHTNESS_OFF_FLOAT &&
-                                        mButtonTimeout != 0) {
-                                    nextTimeout = now + mButtonTimeout;
+                                if ((!mButtonLightOnKeypressOnly || mButtonPressed) &&
+                                        !mProximityPositive) {
+                                    mButtonsLight.setBrightness(buttonBrightness);
+                                    mButtonPressed = false;
+                                    if (buttonBrightness != PowerManager.BRIGHTNESS_OFF_FLOAT &&
+                                            mButtonTimeout != 0) {
+                                        mButtonOn = true;
+                                        if (now + mButtonTimeout < nextTimeout) {
+                                            nextTimeout = now + mButtonTimeout;
+                                        }
+                                    }
+                                } else if (mButtonLightOnKeypressOnly && mButtonOn &&
+                                        mLastButtonActivityTime + mButtonTimeout < nextTimeout) {
+                                    nextTimeout = mLastButtonActivityTime + mButtonTimeout;
                                 }
                             }
                         }
@@ -2349,6 +2379,7 @@ public final class PowerManagerService extends SystemService
                             mUserActivitySummary = USER_ACTIVITY_SCREEN_DIM;
                             if (getWakefulnessLocked() == WAKEFULNESS_AWAKE) {
                                 mButtonsLight.setBrightness(PowerManager.BRIGHTNESS_OFF_FLOAT);
+                                mButtonOn = false;
                             }
                         }
                     }
-- 
2.15.0

